Make ContainsKeyRange return false when start is equal to end

kkaneda opened the pull request:
`ContainsKeyRange` used to return true when a range has `["a", "b")` and `start == end == "b"`. It seems this is not an expected behavior since end key is exclusive.

This commit also fixes `AdminMerge` and test code to pass `(key, nil)` to look up a range with a single key.

kkaneda committed:
Make ContainsKeyRange return false when start is equal to end

ContainsKeyRange used to return true when a range has ["a", "b") and
start == end == "b". It seems this is not an expected behavior since
end key is exclusive.

This commit also fixes AdminMerge() and test code to pass (key, nil) to look up a range with a single key.

petermattis commented:
LGTM, though @spencerkimball might now why the previous behavior existed.

kkaneda commented:
Sure, @spencerkimball , could you take a look? My main motivation was to make the semantics clear before converting the ``rangesByKey` from sorted slice to btree. Without this change, `LookupRange` returns a different result if we change the implement to iterate from the beginning of a slice:

```
for _, rng := range s.rangesByKey {
	if rng.Desc().ContainsKeyRange(startAddr, endAddr) {
		return rng
	}
}
return nil
```

or use btree:

```
// rangesByKey is a btree keyed by ranges' end keys.

var rng *Range
s.rangesByKey.AscendGreaterOrEqual(endAddr, func(i btree.Item) bool {
	rng = (*Range)(i.(*rangeBtreeItem))
	if rng.Desc().ContainsKeyRange(startAddr, endAddr) {
		return false
	}
	rng = nil
	return true
})
return rng
```

spencerkimball commented on storage/range_command.go:
```diff
@@ -961,7 +961,7 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
 		return
 	}
-	subsumedRng := r.rm.LookupRange(desc.EndKey, desc.EndKey)
+	subsumedRng := r.rm.LookupRange(desc.EndKey, desc.EndKey.Next())
```
Instead of adding `.Next()` everywhere we do LookupRange, let's make it support nil as the end key.  You'll need to adjust `storage/engine/keys.go:KeyAddress()` to return nil if nil is supplied.

kkaneda commented:
Thanks, @spencerkimball . Updated the PR as you suggested (`Scan()` still takes `key.Next()`). A test case for `LookupRange(..., nil)` is added to `TestStoreAddRemoveRanges`.

It seems `KeyAddress(nil)` already returns `nil` since `bytes.HasPrefix(nil, ...)` returns false, but I've added an explicit check.

tamird commented:
lgtm

spencerkimball commented:
LGTM
